<a href="https://colab.research.google.com/github/sanskritirk/ProjectWorkfor_DataAnalysisCourse/blob/master/Project5_GenderVoicesClassification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project 5: Gender Voices Classification

### Overview

In this project, you are provided with a dataset which contains some statistical information about the audio frequencies of different male and female voices. Based on the information provided, you have to find out which voice belongs to which gender using the `RandomForestClassifier` algorithm.

The data description is provided below for the dataset.

<img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/project-4/gender_voice_dataset_attributes.png' width=600>

*Dataset credits: `https://www.mldata.io/dataset-details/gender_voice/`*

Please note that

1. In the `label` column, `0` denotes the male class and `1` denotes the female class.

2. You are not required to know the meaning of all the other attributes in the dataset. You just need to deploy a prediction model with these attributes as the feature variables and the `label` column as the target variable. 

You can get the datasets from the following links:

1. Train dataset 
  
  `https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/project-4/gender-voice-train.csv`

2. Test dataset 

  `https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/project-4/gender-voice-test.csv`

---

### Project Requirements

1. Create a pandas DataFrame for the train and test datasets.

2. Display the first five rows of both the training and test DataFrames.

3. Display the last five rows of both the training and test DataFrames.

4. Find the number of rows and columns in the train and test DataFrames.

5. Check for the missing values in the train and test DataFrames.

6. Count the number of `male` and `female` classes in the train DataFrame. 

7. Separate the feature variables, i.e., `x_train` and `x_test` from both the DataFrames.

8. Separate the target variable, i.e., `y_train` and `y_test` from both the DataFrames.

9. Apply the `RandomForestClassifier` machine learning model to predict the `males` and `female` classes in the test DataFrame, i.e, `x_test`.

10. Print the confusion matrix and the classification report to evaluate your prediction model. Also, based on the confusion matrix, precision, recall and f1-score values, report whether the prediction model deployed by you is making accurate predictions or not.

---

#### 1. Create The DataFrames

Create the Pandas DataFrames for both the training and test datasets.

In [ ]:
# Create the DataFrames for both the train and test datasets and store them in the 'train_df' and 'test_df' variables respectively.
import pandas as pd
train_df=pd.read_csv('https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/project-4/gender-voice-train.csv')
test_df=pd.read_csv('https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/project-4/gender-voice-test.csv')

---

#### 2. Display First Five Rows

Display the first five rows of both the `train_df` and `test_df` DataFrames.

In [ ]:
# Display the first 5 rows of the 'train_df' DataFrame.
train_df.head()

label  meanfreq        sd  ...     maxdom    dfrange   modindx
0      1  0.235668  0.035568  ...   9.140625   9.117188  0.087997
1      0  0.190578  0.059717  ...   4.757812   4.734375  0.194908
2      0  0.202016  0.062551  ...   8.460938   8.437500  0.085771
3      1  0.236232  0.041141  ...  10.617188  10.593750  0.103917
4      0  0.189003  0.061455  ...   5.953125   5.929688  0.119684

[5 rows x 21 columns]

In [ ]:
# Display the first 5 rows of the 'test_df' DataFrame.
test_df.head()

label  meanfreq        sd  ...     maxdom    dfrange   modindx
0      1  0.186833  0.027472  ...   0.226562   0.218750  0.108929
1      0  0.188879  0.060316  ...   8.718750   8.695312  0.098712
2      0  0.150705  0.087127  ...   5.562500   5.554688  0.332396
3      1  0.183667  0.040607  ...   3.421875   3.414062  0.166503
4      1  0.205159  0.039543  ...  10.804688  10.617188  0.113644

[5 rows x 21 columns]

---

#### 3. Display Last Five Rows

Display the last five rows of both the `train_df` and `test_df` DataFrames.

In [ ]:
# Display the last 5 rows of the 'train_df' DataFrame.
train_df.tail()

label  meanfreq        sd  ...    maxdom   dfrange   modindx
2117      1  0.166738  0.052677  ...  0.214844  0.053711  0.136364
2118      0  0.198718  0.058959  ...  4.968750  4.945312  0.155766
2119      0  0.202333  0.063001  ...  7.429688  7.406250  0.093438
2120      0  0.181799  0.058102  ...  4.359375  4.335938  0.062312
2121      0  0.167732  0.066225  ...  6.023438  6.000000  0.123384

[5 rows x 21 columns]

In [ ]:
# Display the last 5 rows of the 'test_df' DataFrame.
test_df.tail()

label  meanfreq        sd  ...    maxdom   dfrange   modindx
1041      0  0.143504  0.073334  ...  2.335938  2.328125  0.203020
1042      0  0.151532  0.058071  ...  0.776367  0.688477  0.650870
1043      1  0.230144  0.033577  ...  8.648438  8.414062  0.120313
1044      0  0.204982  0.061316  ...  0.796875  0.648438  0.312048
1045      0  0.171713  0.061671  ...  5.789062  5.757812  0.198301

[5 rows x 21 columns]

---

#### 4. Rows & Columns In Train DataFrame

Find the number of rows and columns in both the `train_df` and `test_df` DataFrames.

In [ ]:
# Print the number of rows and columns in both the DataFrames.
print("No of columns and rows in train dataframe:",train_df.shape)
print("No of columns and rows in test dataframe:",test_df.shape)

No of columns and rows in train dataframe: (2122, 21)
No of columns and rows in test dataframe: (1046, 21)


---

#### 5. Check For Missing Values 

Check whether any of the columns in both the `train_df` and `test_df` DataFrames has any missing value.

In [ ]:
# Check for the missing values in the 'train_df' DataFrame.
num_missing_values = 0

for column in train_df.columns:
  for item in train_df[column].isnull():
    if item == True:
      num_missing_values += 1

print("Number of missing values in train dataset:",num_missing_values)

Number of missing values in train dataset: 0


In [ ]:
# Check for the missing values in the 'test_df' DataFrame.
num_missing_values1 = 0

for column in test_df.columns:
  for item in test_df[column].isnull():
    if item == True:
      num_missing_values1 += 1

print("Number of missing values in test dataset:",num_missing_values1)

Number of missing values in test dataset: 0


---

#### 6. Count The `male` & `female` Classes

Find out the number of `male` and `female` values in both the `train_df` and `test_df` DataFrames.

In [ ]:
# Print the count of the 'male' and 'female' classes in the 'train_df' DataFrame.
train_df['label'].value_counts()

1    1085
0    1037
Name: label, dtype: int64

In [ ]:
# Print the count of the 'male' and 'female' classes in the 'test_df' DataFrame.
test_df['label'].value_counts()

0    547
1    499
Name: label, dtype: int64

---

#### 7. Feature Variables Extraction

Get the feature variables, i.e., `x_train` and `x_test` from both the `train_df` and `test_df` DataFrames respectively. Then, display the first 5 rows of `x_train` and `x_test` DataFrames.

In [ ]:
# Get the feature variables from the 'train_df' DataFrame.
x_train=train_df.iloc[:,1:]
# Display the first 5 rows of the 'x_train' DataFrame.
x_train.head()

meanfreq        sd    median  ...     maxdom    dfrange   modindx
0  0.235668  0.035568  0.240065  ...   9.140625   9.117188  0.087997
1  0.190578  0.059717  0.190747  ...   4.757812   4.734375  0.194908
2  0.202016  0.062551  0.233775  ...   8.460938   8.437500  0.085771
3  0.236232  0.041141  0.244934  ...  10.617188  10.593750  0.103917
4  0.189003  0.061455  0.202376  ...   5.953125   5.929688  0.119684

[5 rows x 20 columns]

In [ ]:
# Get the feature variables from the 'test_df' DataFrame.
x_test=test_df.iloc[:,1:]
# Display the first 5 rows of the 'x_test' DataFrame.
x_test.head()

meanfreq        sd    median  ...     maxdom    dfrange   modindx
0  0.186833  0.027472  0.184325  ...   0.226562   0.218750  0.108929
1  0.188879  0.060316  0.195537  ...   8.718750   8.695312  0.098712
2  0.150705  0.087127  0.174299  ...   5.562500   5.554688  0.332396
3  0.183667  0.040607  0.182534  ...   3.421875   3.414062  0.166503
4  0.205159  0.039543  0.210805  ...  10.804688  10.617188  0.113644

[5 rows x 20 columns]

---

#### 8. Target Variable Extraction

Get the target variables, i.e., `y_train` and `y_test` from both the `train_df` and `test_df` DataFrames respectively. Then, display the first 5 rows of `y_train` and `y_test` Pandas series.

In [ ]:
# Get the target variable from the 'train_df' DataFrame.
y_train=train_df.iloc[:,0]
# Display the first 5 rows of the 'y_train' Pandas series.
y_train.head()

0    1
1    0
2    0
3    1
4    0
Name: label, dtype: int64

In [ ]:
# Get the target variable from the 'test_df' DataFrame.
y_test=test_df.iloc[:,0]
# Display the first 5 rows of the 'y_test' Pandas series.
y_test.head()

0    1
1    0
2    0
3    1
4    1
Name: label, dtype: int64

--- 

#### 9. Building A Random Forest Classifier Model

In [ ]:
# Build a Random Forest Classifier model.
# Import the 'RandomForestClassifier' module.
from sklearn.ensemble import RandomForestClassifier
# Import the confusion_matrix and classification_report modules.
from sklearn.metrics import confusion_matrix,classification_report

In [ ]:
# Predict the target variable based on the feature variables of the test DataFrame.
rf_clf = RandomForestClassifier(n_jobs=-1, n_estimators=50)
rf_clf.fit(x_train, y_train)

print(rf_clf.score(x_train, y_train))
y_predicted = rf_clf.predict(x_test)
print("y predicted:",y_predicted)

y_predicted = pd.Series(y_predicted)
print(y_predicted.head())

print(y_predicted.value_counts())

1.0
y predicted: [1 0 0 ... 1 0 1]
0    1
1    0
2    0
3    1
4    1
dtype: int64
0    537
1    509
dtype: int64


---

#### 10. Confusion Matrix & Classification Report

Print the confusion matrix and classification report to evaluate the model. Interpret and report the results obtained from the confusion matrix and the classification report.

In [ ]:
# Print the confusion matrix to see the number of TN, FN, TP and FP values.
confusion_matrix(y_test,y_predicted)

array([[530,  17],
       [  7, 492]])

In [ ]:
# Print the precision, recall and f1-score values for both the male and female classes.
print(classification_report(y_test,y_predicted))

              precision    recall  f1-score   support

           0       0.99      0.97      0.98       547
           1       0.97      0.99      0.98       499

    accuracy                           0.98      1046
   macro avg       0.98      0.98      0.98      1046
weighted avg       0.98      0.98      0.98      1046



**Write your interpretation of the results here.**

- Interpretation 1: values predicted as class 1 and are actually class 1=530

- Interpretation 2: values predicted as class 1 but actually belong to  class 0=7

- Interpretation 3:values predicted as class 0 and actually belong to  class 0=17
  
           values predicted as class 0 but actually belong to  class 1=492

  $\dots$

- Interpretation N


---

### How To Solve & Submit The Project

Follow the steps described below to solve the project and then submit it.

1. Click on the link provided below to open the Colab file for this project.
   
   https://colab.research.google.com/drive/1eP0L1-vjjYnZSHtiEQ1EL0UFQ3NiyA65

2. Create the duplicate copy of the Colab file. Here are the steps to create the duplicate copy:

    - Click on the **File** menu. A new drop-down list will appear.

      <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lesson-0/0_file_menu.png' width=500>

    - Click on the **Save a copy in Drive** option. A duplicate copy will get created. It will open up in the new tab on your web browser.

      <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lesson-0/1_create_colab_duplicate_copy.png' width=500>

     - After creating the duplicate copy of the notebook, please rename it in the **YYYY-MM-DD_StudentName_Project5** format. 

3. Now, write your code in the prescribed code cells.

4. After finishing the project, click on the **Share** button on the top right corner of the notebook. A new dialog box will appear.

  <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lesson-0/2_share_button.png' width=500>

5. In the dialog box, click on the **Copy link** button.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-1/1_copy_link.png' width=500>

6. The link of the duplicate copy (named as **YYYY-MM-DD_StudentName_Project5**) of the notebook will get copied which you can paste in the submission box on the WhiteHat Jr project submission interface.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-1/2_copy_link_confirmation.png' width=500>

---